File: cpp/queue.cpp

#include <iostream>
using namespace std;

struct Node
{
    int val;
    Node* next;
    Node(int x) : val(x), next(nullptr) {}
};

struct Queue
{
    Node* front;
    Node* rear;
    Queue() : front(nullptr), rear(nullptr) {}

    bool empty()
    {
        return front == nullptr;
    }

    void enqueue(int x)
    {
        Node* node = new Node(x);
        if (!rear)
        {
            front = rear = node;
        }    
        else
        {
            rear->next = node;
            rear = node;
        }
    }

    int dequeu()
    {
        if (!front)
            return 0;
        Node* t = front;
        int a = t->val;
        front = front->next;
        if (!front)
            rear = nullptr;
        delete t;
        return a;
    }

    void printAll()
    {
        if (empty())
        {
            std::cout << "queue empty\n";
            return;
        }
        std::cout << front->val << " <- front\n";
        Node* curr = front->next;
        while (curr)
        {
            std::cout << curr->val;
            std::cout << (curr->next != nullptr ? "\n" : " <- rear\n");
            curr = curr->next;
        }
    }

    int peek()
    {
        if (empty())
            return 0;
        return front->val;
    }


    ~Queue()
    {
        while (!empty())
            dequeu();
    }

};


int main()
{
    Queue line;

    line.enqueue(400);
    line.printAll();
    std::cout << std::endl;
    
    line.enqueue(600);
    line.printAll();
    std::cout << std::endl;
    
    line.enqueue(800);
    line.printAll();
    std::cout << std::endl;
    
    line.dequeu();
    line.printAll();
    std::cout << std::endl;
    
    line.dequeu();
    line.printAll();
    
    line.dequeu();
    line.printAll();
}

File: python/queue.py
class Node:
    def __init__(self, x):
        self.val = x
        self.next = None


class Queue:
    def __init__(self):
        self.front = None
        self.rear = None

    def empty(self):
        return self.front is None

    def enqueue(self, x):
        node = Node(x)
        if not self.rear:
            self.front = self.rear = node
        else:
            self.rear.next = node
            self.rear = node

    def dequeue(self):
        if self.empty():
            print("Queue is empty")
            return
        tmp = self.front
        val = tmp.val
        self.front = tmp.next
        if not self.front:
            self.rear = None
        return val

    def peek(self):
        if self.empty():
            return None
        return self.front.val

    def print_all(self):
        if self.empty():
            print("Queue empty")
            return

        print(f"{self.front.val} <- front")

        curr = self.front.next
        while curr:
            if not curr.next:
                print(f"{curr.val} <- rear")
            else:
                print(curr.val)
            curr = curr.next


line = Queue()

line.enqueue(400)
line.print_all()
print()

line.enqueue(500)
line.print_all()
print()

line.enqueue(600)
line.print_all()
print()

line.enqueue(700)
line.print_all()
print()

x = line.dequeue()
line.print_all()
print()
